examples/nlp/data_parallel_training_with_keras_nlp.py

```diff
@@ -0,0 +1,221 @@
+used for pretraining the BERT model. We will filter out short lines to ensure that the
+data has enough context for training.
+"""
+
+keras.utils.get_file(
+    origin="https://s3.amazonaws.com/research.metamind.io/wikitext/wikitext-2-v1.zip",
+    extract=True,
+)
+wiki_dir = os.path.expanduser("~/.keras/datasets/wikitext-2/")
+
+# Load wikitext-103 and filter out short lines.
+wiki_train_ds = (
+    tf.data.TextLineDataset(wiki_dir + "wiki.train.tokens")
+    .filter(lambda x: tf.strings.length(x) > 100)
+    .batch(PRETRAINING_BATCH_SIZE)
+    .cache()
+    .prefetch(tf.data.AUTOTUNE)
+)
+wiki_val_ds = (
+    tf.data.TextLineDataset(wiki_dir + "wiki.valid.tokens")
+    .filter(lambda x: tf.strings.length(x) > 100)
+    .batch(PRETRAINING_BATCH_SIZE)
+    .cache()
+    .prefetch(tf.data.AUTOTUNE)
+)
+wiki_test_ds = (
+    tf.data.TextLineDataset(wiki_dir + "wiki.test.tokens")
+    .filter(lambda x: tf.strings.length(x) > 100)
+    .batch(PRETRAINING_BATCH_SIZE)
+    .cache()
+    .prefetch(tf.data.AUTOTUNE)
+)
+
+"""
+In the above code, we download the wikitext-2 dataset and extract it. Then, we define
+three datasets: wiki_train_ds, wiki_val_ds, and wiki_test_ds. These datasets are
+filtered to remove short lines and are batched for efficient training.
+"""
+
+EPOCHS = 3
+
+"""
+It's a common practice to use a decayed learning rate in NLP training/tuning. We'll
+use `PolynomialDecay` schedule here.
+"""
+
+total_training_steps = sum(1 for _ in wiki_train_ds.as_numpy_iterator()) * EPOCHS
+lr_schedule = tf.keras.optimizers.schedules.PolynomialDecay(
+    5e-5,
+    decay_steps=total_training_steps,
+    end_learning_rate=0.0,
+)
+
+
+class PrintLR(tf.keras.callbacks.Callback):
+    def on_epoch_end(self, epoch, logs=None):
+        print(
+            f"\nLearning rate for epoch {epoch + 1} is {model_dist.optimizer.lr.numpy()}"
+        )
+
+
+"""
+Let's also make a callback to TensorBoard, this will enable visualization of different
+metrics while we train the model in later part of this tutorial We put all the callbacks
+together as follows:
+"""
+callbacks = [
+    tf.keras.callbacks.TensorBoard(log_dir="./logs"),
+    PrintLR(),
+]
+
+
+print(tf.config.list_physical_devices("GPU"))
+
+"""
+To do single-host, multi-device synchronous training with a Keras model, you would use
+the `tf.distribute.MirroredStrategy` API. Here's how it works:
+
+- Instantiate a `MirroredStrategy`, optionally configuring which specific devices you
+want to use (by default the strategy will use all GPUs available).
+- Use the strategy object to open a scope, and within this scope, create all the Keras
+objects you need that contain variables. Typically, that means **creating & compiling the
+model** inside the distribution scope.
+- Train the model via `fit()` as usual.
+"""
+
+strategy = tf.distribute.MirroredStrategy()
+print("Number of devices: {}".format(strategy.num_replicas_in_sync))
```

Comment:
.format() is generally not recommended in Python 3.x. You can use f-string here.

```
print(f"Number of devices: {strategy.num_replicas_in_sync}")
```

Reply:
Done !